### PR TITLE
Cloud blob store changes for serving data

### PR DESCRIPTION
**Pull request by ankagrawal**
This PR makes CloudBlobStore compatible with BlobStore.
Also makes changes in CloudBlobStore with respect to validations and functionality required for serving data.

### CODE REVIEW

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobMetadata.java [hunk 1]
```diff
@@ -435,8 +435,36 @@ public void setLifeVersion(short lifeVersion) {
...
+  public boolean isUndeleted() {
+    return !isDeleted() && lifeVersion > 0;
+  }
+
+  /**
+   * @param retentionPeriod period for which blobs marked a deleted aren't compacted away.
+   * @return true if deletion time is outside retention window. false otherwise.
+   */
+  public boolean isCompactionCandidate(long retentionPeriod) {
+    return isDeleted() && deletionTime <= (System.currentTimeMillis() - retentionPeriod);
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobMetadata.java:467]: This misses the expired blobs.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudDestination.java [hunk 2]
```diff
@@ -53,28 +53,34 @@ boolean uploadBlob(BlobId blobId, long inputLength, CloudBlobMetadata cloudBlobM
    * @param blobId id of the Ambry blob
    * @param deletionTime time of blob deletion
    * @param lifeVersion life version of the blob to be deleted.
+   * @param cloudUpdateValidator {@link CloudUpdateValidator} object passed by called to validate the update.
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/CloudDestination.java:56]: caller

Comment (lightningrob) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/CloudDestination.java:56]: And validate the delete.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudDestination.java [hunk 3]
```diff
@@ -53,28 +53,34 @@ boolean uploadBlob(BlobId blobId, long inputLength, CloudBlobMetadata cloudBlobM
...
    */
-  boolean deleteBlob(BlobId blobId, long deletionTime, short lifeVersion) throws CloudStorageException;
+  boolean deleteBlob(BlobId blobId, long deletionTime, short lifeVersion, CloudUpdateValidator cloudUpdateValidator)
+      throws CloudStorageException;
 
   /**
    * Undelete the blob from cloud destination, and update the new life version.
    * @param blobId id of the Ambry blob.
    * @param lifeVersion new life version to update.
+   * @param cloudUpdateValidator {@link CloudUpdateValidator} object passed by called to validate the update.
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/CloudDestination.java:67]: caller to validate the undelete

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudDestination.java [hunk 4]
```diff
@@ -53,28 +53,34 @@ boolean uploadBlob(BlobId blobId, long inputLength, CloudBlobMetadata cloudBlobM
...
-  short undeleteBlob(BlobId blobId, short lifeVersion) throws CloudStorageException;
+  short undeleteBlob(BlobId blobId, short lifeVersion, CloudUpdateValidator cloudUpdateValidator)
+      throws CloudStorageException;
 
   /**
    * Update expiration time of blob in the cloud destination.
    * @param blobId id of the Ambry blob
    * @param expirationTime the new expiration time
-   * @return flag indicating whether the blob was updated
+   * @param cloudUpdateValidator {@link CloudUpdateValidator} object passed by called to validate the update.
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/CloudDestination.java:78]: caller

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudUpdateValidator.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+import java.util.Map;
+
+
+/**
+ * Interface for Ambry validation logic for updates requested from cloud destination.
+ */
+public interface CloudUpdateValidator {
+  /**
+   * Validate operation on {@link CloudBlobMetadata} in cloud destination for the operation on blob with
+   * given {@link StoreKey} and new requested life version.
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/CloudUpdateValidator.java:27]: new requested life version doesn't appear to be an argument.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudUpdateValidator.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+   * @param metadata {@link CloudBlobMetadata} object obtained from cloud destination.
+   * @param key {@link StoreKey} of the blob being updated.
+   * @param updateFields {@link Map} of fields and new values requested for update.
+   * @throws StoreException if validation fails.
+   */
+  void validateUpdate(CloudBlobMetadata metadata, StoreKey key, Map<String, Object> updateFields) throws StoreException;
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/CloudUpdateValidator.java:33]: It seems strange that a class in cloud package is throwing StoreException.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudUpdateValidator.java (hunk 5 above)
Comment (lightningrob) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/CloudUpdateValidator.java:27]: I'd like to see clearer description of what is validated, or what would cause validation to fail.  And specifically that you are validating application of update fields to existing metadata.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java [hunk 7]
```diff
@@ -391,11 +377,15 @@ private boolean deleteIfNeeded(BlobId blobId, long deletionTime, short lifeVersi
     // Note: always check cache before operation attempt, since this could be a retry after a CONFLICT error,
     // in which case the cache may have been updated by another thread.
     if (!checkCacheState(blobKey, lifeVersion, BlobState.DELETED)) {
-      boolean deleted = cloudDestination.deleteBlob(blobId, deletionTime, lifeVersion);
+      boolean deleted = cloudDestination.deleteBlob(blobId, deletionTime, lifeVersion,
+          (metadata, key, updateFields) -> preDeleteValidation(metadata, key, updateFields));
       addToCache(blobKey, lifeVersion, BlobState.DELETED);
       return deleted;
+    } else {
+      throw new CloudStorageException("Error updating blob metadata",
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java:385]: We might need to distinguish between vcr and live traffic.  For VCR, shouldUpload() can return false if the TTL is low.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java [hunk 8]
```diff
@@ -314,13 +290,20 @@ private void putBlob(MessageInfo messageInfo, ByteBuffer messageBuf, long size)
...
       }
       addToCache(blobId.getID(), (short) 0, BlobState.CREATED);
+      if (!uploaded) {
+        throw new StoreException(String.format("Another blob with same key %s exists in store", blobId.getID()),
+            StoreErrorCodes.Already_Exist);
+      }
     } else {
-      logger.trace("Blob is skipped: {}", messageInfo);
       vcrMetrics.blobUploadSkippedCount.inc();
+      throw new CloudStorageException("Error updating blob metadata", new StoreException(
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java:304]: Why not just throw a StoreException here?
Also, for VCR case it may not be an error (that's why we mark it skipped).

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudDestination.java (hunk 2 above)
Comment (ankagrawal) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/CloudDestination.java:56]: fixed.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudDestination.java (hunk 3 above)
Comment (ankagrawal) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/CloudDestination.java:67]: fixed.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudDestination.java (hunk 4 above)
Comment (ankagrawal) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/CloudDestination.java:78]: fixed.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudUpdateValidator.java (hunk 5 above)
Comment (ankagrawal) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/CloudUpdateValidator.java:27]: Added comments to capture the description, while keeping it relatively generic so that it can potentially be applied to future CloudDestination extensions. Let me know if it looks good.

**Comment by ankagrawal** [comment]
> I think many of the places you added exception throws for live traffic will break the VCR. The tests don't seem to capture that since a dummy validator is passed to AzureCloudDestination. (We don't really have a CloudBlobStore integration test which is a current gap.)

After closer review, fixed the cases where validations could break VCR.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java [hunk 9]
```diff
@@ -295,15 +295,23 @@ private void putBlob(MessageInfo messageInfo, ByteBuffer messageBuf, long size)
...
             StoreErrorCodes.Already_Exist);
       }
     } else {
       vcrMetrics.blobUploadSkippedCount.inc();
-      throw new CloudStorageException("Error updating blob metadata", new StoreException(
-          String.format("Another blob with same key %s exists in store", messageInfo.getStoreKey().getID()),
-          StoreErrorCodes.Already_Exist));
+      // The only case where its ok to see a put request for a already seen blob is, during replication if the blob is
+      // expiring within {@link CloudConfig#vcrMinTtlDays} for vcr to upload.
+      if (isVcr && !isExpiringSoon(messageInfo)) {
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java:310]: I don't understand this logic.  What about the case where we have reset the replica tokens but not deleted the blob store?  The put needs to be idempotent for VCR.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudUpdateValidator.java [hunk 10]
```diff
@@ -23,12 +23,15 @@
...
-   * Validate operation on {@link CloudBlobMetadata} in cloud destination for the operation on blob with
-   * given {@link StoreKey} and new requested life version.
+   * Validate the sanity of update operation on given {@code updateFields} against existing {@link CloudBlobMetadata} in
+   * cloud destination for the blob with key {@code key}. This can be used to hook Ambry store related validation logic
+   * during {@link CloudDestination} specific get-check-update flow.
    * @param metadata {@link CloudBlobMetadata} object obtained from cloud destination.
    * @param key {@link StoreKey} of the blob being updated.
    * @param updateFields {@link Map} of fields and new values requested for update.
+   * @return if no exception needs to thrown, return false if the operation is not needed (maybe cloud state is more
+   * recent). true otherwise.
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/CloudUpdateValidator.java:33]: Rephrase as:
@return true if the operation can proceed, or false if it is not needed (e.g. cloud state is more recent).
The exception part is implicit.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobMetadata.java [hunk 11]
```diff
@@ -432,11 +432,32 @@ public void setLifeVersion(short lifeVersion) {
...
+  public boolean isUndeleted() {
+    return !isDeleted() && lifeVersion > 0;
+  }
+
+  /**
+   * @param retentionPeriod period for which blobs marked a deleted aren't compacted away.
+   * @return true if deletion time is outside retention window or blob is expired. false otherwise.
+   */
+  public boolean isCompactionCandidate(long retentionPeriod) {
+    return (isDeleted() && deletionTime <= (System.currentTimeMillis() - retentionPeriod)) || isExpired();
```
[the remark this answers is not in the document]
Comment (ankagrawal) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobMetadata.java:460]: ok.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java [hunk 12]
```diff
@@ -295,15 +295,23 @@ private void putBlob(MessageInfo messageInfo, ByteBuffer messageBuf, long size)
...
+      if (!uploaded && !isVcr) {
+        // If put is coming from frontend, then uploadBlob must be true. Its not acceptable that a blob already exists.
+        // If put is coming from vcr, then findMissingKeys might have reported a key to be missing even though the blob
+        // was uploaded to ABS. This can happen, if previously, the upload to ABS succeeded but the upload couldn't make
+        // it to cosmos.
         throw new StoreException(String.format("Another blob with same key %s exists in store", blobId.getID()),
             StoreErrorCodes.Already_Exist);
       }
     } else {
       vcrMetrics.blobUploadSkippedCount.inc();
```
[the remark this answers is not in the document]
Comment (ankagrawal) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java:307]: I guess this (line 307) is the else case for shouldUpload() check. So should be fine.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java (hunk 9 above)
Comment (ankagrawal) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java:310]: I guess one way to look at this is that for vcr (replication), put happens when ReplicaThread has made the determination that this blob is not present locally by calling the CloudBlobStore::findMissingKeys(). So in case where we have reset the replica tokens but not deleted the blob store, a call to findMissingKeys should say that this blob is indeed present, and Replica Thread should not attempt a put() for this blob.

Let me know if I am missing something here.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java [hunk 13]
```diff
@@ -522,12 +530,20 @@ private void validateAccountAndContainer(Map<String, CloudBlobMetadata> cloudBlo
...
    */
-  private void preDeleteValidation(CloudBlobMetadata metadata, StoreKey key, Map<String, Object> updateFields)
+  private boolean preDeleteValidation(CloudBlobMetadata metadata, StoreKey key, Map<String, Object> updateFields)
       throws StoreException {
     validateAccountAndContainer(Collections.singletonMap(metadata.getId(), metadata), Collections.singletonList(key));
     short requestedLifeVersion = (short) updateFields.get(FIELD_LIFE_VERSION);
+    if (isVcr) {
+      // This is a delete request from vcr. Apply delete only if incoming life version is more recent. Don't throw any
+      // exception because replication relies on findMissingKeys which in turn relies on cosmos state for CloudBlobStore.
+      // Cosmos could be missing some updates that were applied to ABS.
```
[the remark this answers is not in the document]
Comment (ankagrawal) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java:543]: ok.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java [hunk 14]
```diff
@@ -554,12 +563,19 @@ private void preDeleteValidation(CloudBlobMetadata metadata, StoreKey key, Map<S
...
    * @throws StoreException if validation fails.
    */
-  private void preTtlUpdateValidation(CloudBlobMetadata metadata, StoreKey key, Map<String, Object> updateFields)
+  private boolean preTtlUpdateValidation(CloudBlobMetadata metadata, StoreKey key, Map<String, Object> updateFields)
       throws StoreException {
     validateAccountAndContainer(Collections.singletonMap(metadata.getId(), metadata), Collections.singletonList(key));
     long now = System.currentTimeMillis();
+    if (isVcr) {
+      // For vcr don't update ttl if already updated. Don't throw any exception because replication relies on
+      // findMissingKeys which in turn relies on cosmos state for CloudBlobStore. Cosmos could be missing some updates
```
[the remark this answers is not in the document]
Comment (ankagrawal) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java:575]: ok

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java [hunk 15]
```diff
@@ -579,10 +596,16 @@ private void preTtlUpdateValidation(CloudBlobMetadata metadata, StoreKey key, Ma
    * @param updateFields {@link Map} of fields and values being updated.
    * @throws StoreException if validation fails.
    */
-  private void preUndeleteValidation(CloudBlobMetadata metadata, StoreKey key, Map<String, Object> updateFields)
+  private boolean preUndeleteValidation(CloudBlobMetadata metadata, StoreKey key, Map<String, Object> updateFields)
```
[the remark this answers is not in the document]
Comment (ankagrawal) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java:599]: ok

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java [hunk 16]
```diff
@@ -579,10 +596,16 @@ private void preTtlUpdateValidation(CloudBlobMetadata metadata, StoreKey key, Ma
...
       throws StoreException {
     validateAccountAndContainer(Collections.singletonMap(metadata.getId(), metadata), Collections.singletonList(key));
     short requestedLifeVersion = (short) updateFields.get(FIELD_LIFE_VERSION);
+    if (isVcr) {
+      // This is an undelete request from vcr. Apply undelete only if incoming life version is more recent. Don't throw
+      // any exception because replication relies on findMissingKeys which in turn relies on cosmos state for CloudBlobStore.
+      // Cosmos could be missing some updates that were applied to ABS.
```
[the remark this answers is not in the document]
Comment (ankagrawal) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java:606]: ok

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudUpdateValidator.java (hunk 10 above)
Comment (ankagrawal) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/CloudUpdateValidator.java:33]: ok

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java [hunk 17]
```diff
@@ -419,32 +417,34 @@ public short undelete(MessageInfo info) throws StoreException {
    * @throws CloudStorageException in case any exception happens during undelete.
    */
   private short undeleteIfNeeded(BlobId blobId, short lifeVersion) throws CloudStorageException {
-    // TODO: Currently this is implemented for undeletes via replication only for DR.
     // See note in deleteIfNeeded.
     if (!checkCacheState(blobId.getID(), lifeVersion, BlobState.CREATED)) {
-      short newLifeVersion = cloudDestination.undeleteBlob(blobId, lifeVersion);
+      short newLifeVersion = cloudDestination.undeleteBlob(blobId, lifeVersion,
+          (metadata, key, updateFields) -> preUndeleteValidation(metadata, key, updateFields));
```
Comment (cgtz) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java:423]: could you simplify the lambda to `this::preUndeleteValidation`? Same in other places where this method is used.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java [hunk 18]
```diff
@@ -469,17 +469,171 @@ public void updateTtl(List<MessageInfo> infos) throws StoreException {
...
+  /**
+   * Gets the operation time for a blob from blob metadata based on the blob's current state and timestamp recorded for that state.
+   * @param metadata blob metadata from which to derive operation time.
+   * @return operation time.
+   */
+  private long getOperationTime(CloudBlobMetadata metadata) {
+    if (isBlobDeleted(metadata)) {
+      return metadata.getDeletionTime();
+    }
+    return (metadata.getCreationTime() == Utils.Infinite_Time) ? metadata.getUploadTime() : metadata.getCreationTime();
```
Comment (cgtz) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java:634]: does undelete change the operation time?

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java [hunk 19]
```diff
@@ -811,6 +990,8 @@ public void appendFrom(ReadableByteChannel channel, long size) throws StoreExcep
         messageBuf.flip();
         cloudBlobStore.putBlob(messageInfo, messageBuf, size);
         messageIndex++;
+      } catch (StoreException se) {
```
Comment (cgtz) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java:993]: why is this clause added? seems like it is immediately rethrowing the same exc.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudUpdateValidator.java [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+ * Interface for Ambry validation logic for updates requested from cloud destination.
+ */
+public interface CloudUpdateValidator {
+  /**
+   * Validate the sanity of update operation on given {@code updateFields} against existing {@link CloudBlobMetadata} in
+   * cloud destination for the blob with key {@code key}. This can be used to hook Ambry store related validation logic
+   * during {@link CloudDestination} specific get-check-update flow.
+   * @param metadata {@link CloudBlobMetadata} object obtained from cloud destination.
+   * @param key {@link StoreKey} of the blob being updated.
+   * @param updateFields {@link Map} of fields and new values requested for update.
```
Comment (cgtz) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/CloudUpdateValidator.java:31]: note that  the validator may change fields within this map.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudDestination.java [hunk 21]
```diff
@@ -473,7 +483,7 @@ private CloudStorageException toCloudStorageException(String message, Exception
       statusCode = StatusCodes.INTERNAL_SERVER_ERROR;
     }
     // Everything is retryable except NOT_FOUND
-    boolean isRetryable = (statusCode != StatusCodes.NOTFOUND);
+    boolean isRetryable = (statusCode != StatusCodes.NOTFOUND && !(e instanceof StoreException));
```
Comment (cgtz) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudDestination.java:486]: What is the case where a `StoreException` gets here? Is this when the CloudUpdateValidator throws?

File: ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudDestination.java [hunk 22]
```diff
@@ -293,21 +306,18 @@ private boolean updateBlobMetadata(BlobId blobId, Map<String, Object> updateFiel
           // attempt fails.  So we check for that case here.
           CloudBlobMetadata cosmosMetadata = cosmosDataAccessor.getMetadataOrNull(blobId);
           if (cosmosMetadata != null) {
-            if (cosmosMetadata.isDeletedOrExpired()) {
+            if (cosmosMetadata.isCompactionCandidate(
+                TimeUnit.HOURS.toMillis(cloudConfig.cloudBlobCompactionIntervalHours))) {
               logger.warn("Inconsistency: Cosmos contains record for inactive blob {}, removing it.", blobId.getID());
               cosmosDataAccessor.deleteMetadata(cosmosMetadata);
               azureMetrics.blobUpdateRecoverCount.inc();
```
Comment (cgtz) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudDestination.java:313]: It seems like the code will now throw in this case. Is there a reason for the change?

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudStorageManager.java [hunk 23]
```diff
@@ -16,6 +16,7 @@
 import com.github.ambry.clustermap.ClusterMap;
 import com.github.ambry.clustermap.PartitionId;
 import com.github.ambry.clustermap.ReplicaId;
+import com.github.ambry.config.StoreConfig;
```
Comment (cgtz) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/CloudStorageManager.java:19]: unused import?

File: ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudDestination.java (hunk 21 above)
Comment (ankagrawal) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudDestination.java:486]: yes.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java (hunk 9 above)
Comment (lightningrob) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java:309]: Okay, you may be right.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudDestination.java (hunk 22 above)
Comment (cgtz) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudDestination.java:314]: Thanks for the explanation. I think that makes sense

**Comment by cgtz** [comment]
I can merge this after the conflicts are resolved.

File: ambry-cloud/src/test/java/com/github/ambry/cloud/azure/AzureIntegrationTest.java [hunk 24]
```diff
@@ -502,8 +502,12 @@ public void testRepairAfterIncompleteCompaction() throws Exception {
     azureDest.getAzureBlobDataAccessor().purgeBlobs(Collections.singletonList(cloudBlobMetadata));
 
     // Try to delete again (to trigger recovery), verify removed from Cosmos
-    assertFalse("Expected delete to return false",
-        azureDest.deleteBlob(blobId, deletionTime, (short) 0, dummyCloudUpdateValidator));
+    try {
+      assertFalse("Expected delete to return false",
+          azureDest.deleteBlob(blobId, deletionTime, (short) 0, dummyCloudUpdateValidator));
+    } catch (CloudStorageException cex) {
+      assertEquals("Unexpected error code", HttpConstants.StatusCodes.NOTFOUND, cex.getStatusCode());
```
Comment (lightningrob) [inline ambry-cloud/src/test/java/com/github/ambry/cloud/azure/AzureIntegrationTest.java:509]: This doesn't make sense, you can't expect the method to return false and throw an exception.

File: ambry-cloud/src/test/java/com/github/ambry/cloud/CloudTestUtil.java [hunk 25]
```diff
@@ -0,0 +1,116 @@
...
+import com.github.ambry.store.MessageInfo;
+import com.github.ambry.store.MockMessageWriteSet;
+import com.github.ambry.utils.TestUtils;
+import java.nio.ByteBuffer;
+import java.util.Random;
+
+import static com.github.ambry.commons.BlobId.*;
+
+
+public class CloudTestUtil {
```
Comment (lightningrob) [inline ambry-cloud/src/test/java/com/github/ambry/cloud/CloudTestUtil.java:27]: The methods in this class seem unrelated to anything cloud.

File: ambry-cloud/src/test/java/com/github/ambry/cloud/CloudBlobStoreIntegrationTest.java [hunk 26]
```diff
@@ -0,0 +1,515 @@
...
+import static org.junit.Assert.*;
+
+
+/**
+ * Integration Test cases for {@link CloudBlobStore}
+ * Must supply file azure-test.properties in classpath with valid config property values.
+ */
+//@Ignore
+@RunWith(Parameterized.class)
+public class CloudBlobStoreIntegrationTest {
```
Comment (lightningrob) [inline ambry-cloud/src/test/java/com/github/ambry/cloud/CloudBlobStoreIntegrationTest.java:75]: Cool, thanks for adding this.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java [hunk 27]
```diff
@@ -416,16 +416,21 @@ public short undelete(MessageInfo info) throws StoreException {
...
       addToCache(blobId.getID(), newLifeVersion, BlobState.CREATED);
       return newLifeVersion;
     } else {
-      throw new CloudStorageException("Error updating blob metadata",
-          new StoreException("Id " + blobId.getID() + " is already undeleted in cloud", StoreErrorCodes.ID_Undeleted));
+      if (lifeVersion > 0) {
+        throw new StoreException("Id " + blobId.getID() + " is already undeleted in cloud",
+            StoreErrorCodes.ID_Undeleted);
+      } else {
+        throw new StoreException("Id " + blobId.getID() + " not deleted yet in cloud", StoreErrorCodes.ID_Not_Deleted);
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java:432]: Is it possible that is has been deleted in the cloud but didn't go through this vcr so the cache entry is stale?

File: ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudDestination.java [hunk 28]
```diff
@@ -195,8 +195,8 @@ public short undeleteBlob(BlobId blobId, short lifeVersion, CloudUpdateValidator
     updateFields.put(CloudBlobMetadata.FIELD_LIFE_VERSION, lifeVersion);
     updateFields.put(CloudBlobMetadata.FIELD_DELETION_TIME, Utils.Infinite_Time);
     UpdateResponse updateResponse = updateBlobMetadata(blobId, updateFields, cloudUpdateValidator);
-    return updateResponse.metadata.containsKey(CloudBlobMetadata.FIELD_LIFE_VERSION) ? 0
-        : Short.parseShort(updateResponse.metadata.get(CloudBlobMetadata.FIELD_LIFE_VERSION));
+    return updateResponse.metadata.containsKey(CloudBlobMetadata.FIELD_LIFE_VERSION) ? Short.parseShort(
+        updateResponse.metadata.get(CloudBlobMetadata.FIELD_LIFE_VERSION)) : 0;
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudDestination.java:199]: I think you can use Map.getOrDefault() here.

File: ambry-cloud/src/test/java/com/github/ambry/cloud/CloudBlobStoreIntegrationTest.java [hunk 29]
```diff
@@ -0,0 +1,515 @@
...
+  private VerifiableProperties verifiableProperties;
+  private CloudBlobStore cloudBlobStore;
+  private CloudDestination cloudDestination;
+  private Random random = new Random();
+  private short accountId = 101;
+  private short containerId = 5;
+  private long operationTime = System.currentTimeMillis();
+  private boolean isVcr;
+  private PartitionId partitionId;
+  // one day retention
```
Comment (lightningrob) [inline ambry-cloud/src/test/java/com/github/ambry/cloud/CloudBlobStoreIntegrationTest.java:86]: Comment doesn't seem to match code.

File: ambry-cloud/src/test/java/com/github/ambry/cloud/CloudBlobStoreIntegrationTest.java [hunk 30]
```diff
@@ -0,0 +1,515 @@
...
+      }
+    } catch (StoreException ex) {
+      assertEquals("Unexpected error code", ex.getErrorCode(), StoreErrorCodes.ID_Deleted);
+    }
+  }
+
+  /**
+   * Cleanup the test partition by deleting all the blobs of the test partition.
+   */
+  private void cleanup() {
```
Comment (lightningrob) [inline ambry-cloud/src/test/java/com/github/ambry/cloud/CloudBlobStoreIntegrationTest.java:490]: Can this be a common utility used by both (and future) integration tests?

File: ambry-cloud/src/test/java/com/github/ambry/cloud/CloudBlobStoreTest.java [hunk 31]
```diff
@@ -873,11 +940,11 @@ public void testPutWithTtl() throws Exception {
    * @return the generated {@link BlobId}.
    */
   private BlobId addBlobToSet(MockMessageWriteSet messageWriteSet, long size, long expiresAtMs, short accountId,
```
Comment (lightningrob) [inline ambry-cloud/src/test/java/com/github/ambry/cloud/CloudBlobStoreTest.java:942]: Shouldn't these methods be removed here?

Comment (ankagrawal) [reply ambry-cloud/src/test/java/com/github/ambry/cloud/CloudBlobStoreTest.java:942]: done.

File: ambry-cloud/src/test/java/com/github/ambry/cloud/azure/AzureIntegrationTest.java (hunk 24 above)
Comment (ankagrawal) [reply ambry-cloud/src/test/java/com/github/ambry/cloud/azure/AzureIntegrationTest.java:509]: fixed.

**Comment by lightningrob** [comment]
Will merge once CloudBlobStoreTest test is fixed.
